### DISCUSSION TITLE
[Avalonia.Android] GPU Hardware Acceleration doesn't work.

### DISCUSSION THREAD

**Discussion by J0nathan550** [AvaloniaUI/Avalonia#17028] [Q&A] [answered]
Hello, I was developing an Android Application using Avalonia but noticed that scrolling or anything else in my android app is very laggy.

I started to debug, and find out that Avalonia doesn't use GPU rendering in Android and lags badly.

Below are videos showing (with red and green) that gpu (hardware acceleration) is getting called and optimizes phone to 60 fps.
In video there is also demo where those Hardware Acceleration calls are not used, and it's likely that the rendering is done only on CPU.
Also there are comparing between Android Studio app (plain application) and Avalonia application.

https://github.com/user-attachments/assets/450efbc0-4ccf-4571-be96-ae8d827695e4

https://github.com/user-attachments/assets/93d9c895-d022-423e-a510-cbacb5b21551

**Comment by emmauss** [comment]
Also, the dev options you enabled to show the gpu behavior is called View Updates. It only shows when a view is drawn. It will always show 1 update when the view is created on real GPU accelerated views, like games and our renderer. Because we render directly to the gpu surface provided by the OS. So yeah, what you did isn't a good metric for gpu behavior.

**Reply by J0nathan550** [reply]
Well, anyway, if to compare android studio app and well Avalonia, we can actually see the perfomance issue in recycling items. Where in recycler view it doesn't lag at scrolling meanwhile in Avalonia it is. :(

**Reply by emmauss** [reply]
WE have heavily limited by the performance of the dotnet android jit. Without a minimum sample from you, we can't confirm if this is an issue in your app or dotnet.

**Reply by J0nathan550** [reply]
As soon as I get back to home I will send you the source code.

**Reply by J0nathan550** [reply]
Okay, I use the CommunityToolKit.MVVM in order to bind a ObservableCollection of songs. Inside of the ListBox that I use to show the results I have a template with the Image, 3 texts. That in Xaml looks like this:

`SearchView.xaml`:
```xaml
		<ListBox Name="ListBoxSongsResult" Grid.Row="2" Background="Transparent" ItemsSource="{Binding SongsResults}" IsVisible="{Binding IsSongsListVisible}" SelectedItem="{Binding SelectedSong, Mode=TwoWay}" Margin="5,5,5,0">
			<ListBox.ItemTemplate>
				<DataTemplate DataType="models:SongModel">
					<Grid ColumnDefinitions="auto, *, auto">
						<asyncImageLoader:AdvancedImage Margin="5,0,0,0" Width="120" Height="64" Source="{Binding Image}" RenderOptions.BitmapInterpolationMode="HighQuality" />
						<Grid Grid.Column="1" Margin="10,0,10,0" RowDefinitions="*, *">
							<TextBlock Text="{Binding Title}" VerticalAlignment="Bottom" TextTrimming="CharacterEllipsis"/>
							<TextBlock Grid.Row="1" Text="{Binding Author}" VerticalAlignment="Top" TextTrimming="CharacterEllipsis"/>
						</Grid>
						<TextBlock Margin="0,0,5,0" VerticalAlignment="Center" HorizontalAlignment="Center" Grid.Column="2" Text="{Binding Duration}"/>
					</Grid>
				</DataTemplate>
			</ListBox.ItemTemplate>
		</ListBox>
```

Code behind (ViewModel) that processes api and sets the visual elements looks like this:

`SearchViewModel.cs`:
```cs
[RelayCommand]
private void Search()
{
    if (string.IsNullOrEmpty(SearchText)) return;

    IsRegularErrorShown = false;
    IsNoInternetErrorShown = false;
    IsNotFoundErrorShown = false;
    IsLoadingIndicatorShowing = true;
    IsSongsListVisible = false;
    ErrorText = "Loading...";

    SongsResults.Clear();
    cancelSearching.Cancel();
    cancelSearching = new();

    try
    {
        Task.Run(async () =>
        {
            try
            {
                foreach (VideoSearchResult video in await youtubeClient.Search.GetVideosAsync(SearchText, cancelSearching.Token).CollectAsync(50))
                {
                    if (video.Duration == null) return;
                    SongsResults.Add(new SongModel() { Title = video.Title, Author = video.Author.ChannelTitle, Duration = TimeConverter.ConvertMsToTime((long)video.Duration.Value.TotalMilliseconds), Id = video.Id, Image = video.Thumbnails[1].Url });
                }
                SaveSystem.UserData.SearchListSongs = SongsResults;
                SaveSystem.UserData.LastQuery = SearchText;
                SaveSystem.Save();
                if (SongsResults.Count == 0)
                {
                    IsSongsListVisible = false;
                    IsLoadingIndicatorShowing = false;
                    IsNotFoundErrorShown = true;
                    IsRegularErrorShown = false;
                    IsNoInternetErrorShown = false;
                    ErrorText = $"No songs were found...";
                    return;
                }
                IsLoadingIndicatorShowing = false;
                IsNotFoundErrorShown = false;
                IsRegularErrorShown = false;
                IsNoInternetErrorShown = false;
                IsSongsListVisible = true;
                ErrorText = string.Empty;
            }
            catch (HttpRequestException)
            {
                IsSongsListVisible = false;
                IsRegularErrorShown = false;
                IsNoInternetErrorShown = true;
                IsNotFoundErrorShown = false;
                IsLoadingIndicatorShowing = false;
                ErrorText = $"Looks like you have problems with internet...";
            }
            catch (WebException) { /* do nothing */}
            catch (TaskCanceledException) { /* do nothing */}
            catch (Exception ex)
            {
                IsSongsListVisible = false;
                IsRegularErrorShown = true;
                IsNoInternetErrorShown = false;
                IsNotFoundErrorShown = false;
                IsLoadingIndicatorShowing = false;
                ErrorText = $"There was some problem in program, try again...";
                Trace.WriteLine("Unknown Exception to fix: \n" + ex.Message + "\n" + ex.StackTrace);
            }
        }, cancelSearching.Token);
    }
    catch { }
}
```
Output looks like this:
![Screenshot_1](https://github.com/user-attachments/assets/68e2d687-7951-4ef8-b1be-c4d24f39aef6)

I don't believe there's an issue with the code I'm running, as it's a straightforward task of assigning data to a ListBox using MVVM. I can confirm that removing images from the ListBox slightly improves the app's performance because it reduces the rendering load. However, this doesn't affect computer performance; it works as expected. The problem arises on Android, where it lags. I’m confident that I’m using the same data display methods as other popular Android apps, which handle 1,000 images in a view without lag. In contrast, Avalonia lags even with just 50 images.

**Reply by emmauss** [reply, marked as answer]
On Android, we are limited by the performance the Mono jit is able to give us. There's is little we can do about it since all dotnet applications on Android are affected by it.

**Reply by J0nathan550** [reply]
Bad to hear about it :( 

Anyways, it is a still good framework no matter what.

**Reply by J0nathan550** [reply]
I hope in the future dotnet team will do something about it...

**Comment by CeSun** [comment]
Hello, I am very interested in this question. Have you tested it on the iOS platform? On the iOS platform NET runtime should switch to NativeAOT, so theoretically performance should be better
